fix(orderBy): Throw exception when orderBy is given a non array-like object.

nhodges opened the pull request:
Throws exception when orderBy is given a non array-like object to sort.

Closes #11255.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/angular/angular.js/11719)

<!-- Reviewable:end -->

nhodges committed:
fix(orderBy): Throw exception when orderBy is given a non array-like object. #11255

petebacondarwin commented on src/ng/filter/orderBy.js:
```diff
@@ -155,7 +155,7 @@
 function orderByFilter($parse) {
   return function(array, sortPredicate, reverseOrder) {
-    if (!(isArrayLike(array))) return array;
+    if (!(isArrayLike(array))) throw minErr('filter')('notarray', 'Expected array but received: {0}', array);
```
Sadly this error code is specific to the `filterFilter` so we have to provide a new one and associated documentation in `docs/content/error` or alternatively update the documentation to make this error code more generic.

fhernandezn committed:
Documentation to orderby when the parameter is not an array.

frankweb commented:
@nhodges @petebacondarwin Hi guys, I've been working in the documentation needed for this pull request. Please look at this 

https://github.com/angular/angular.js/compare/master...frankweb:orderby-notarray-documentation

I'd like to contribute to this pull request by providing documentation.

nhodges commented:
:+1: @frankweb for providing documentation, I updated my PR to be fast forwardable.

frankweb commented:
@nhodges :+1: do you want to include my documentation in your PR?

nhodges commented:
@frankweb You got it :)

nhodges commented:
@frankweb looks like the task to build the docs is failing after merging, can you take a look?

frankweb commented:
@nhodges Of course, I'll take a look in a while :)

frankweb commented on src/ng/filter/orderBy.js:
```diff
@@ -177,7 +177,7 @@ orderByFilter.$inject = ['$parse'];
   return function(array, sortPredicate, reverseOrder) {
 
-    if (!(isArrayLike(array))) return array;
+    if (!(isArrayLike(array))) throw minErr('filter')('notarray', 'Expected array but received: {0}', array);
```
`grunt test` runs without errors if I change `minErr('filter')` to `minErr('orderBy')`

frankweb commented on test/ng/filter/orderBySpec.js:
```diff
@@ -8,6 +8,11 @@ describe('Filter: orderBy', function() {
   describe('(Arrays)', function() {
+    it('should throw an exception if no array-like object is provided', function() {
+      expect(function() { orderBy({}); }).
+        toThrowMinErr('filter', 'notarray', 'Expected array but received: {}');
```
The same for this line, we need to change `'filter` to `orderBy`

gkalpak commented on docs/content/error/orderBy/notarray.ngdoc:
```diff
@@ -0,0 +1,53 @@
+@fullName Not an array
+@description
+
+This error occurs when {@link ng.orderBy filter} is not used with an array:
```
_filter_  -->  _orderBy_

gkalpak commented on docs/content/error/orderBy/notarray.ngdoc:
```diff
@@ -0,0 +1,53 @@
+             },
+    };
+
+    //if using underscore
```
I don't think we want to have an example using underscore (or any 3rd-party library).

petebacondarwin replied:
I agree. Sorry

frankweb replied:
Ok guys, no problem, do you have any other comment?

frankweb commented on docs/content/error/orderBy/notarray.ngdoc:
```diff
@@ -0,0 +1,53 @@
+    };
+
+    //if using underscore
+    $scope.objToArray = = _.values(obj);
```
We can remove the comment and change 

``` javascript
$scope.objToArray = _.values(obj);
```

to

``` javascript
$scope.objToArray = Object.keys(obj).map(function(a){
   return obj[a] ;
});
```

Doing this we don't need underscore. What do you think @petebacondarwin @gkalpak ?

nhodges committed:
fix(orderBy): Change filter to orderBy in error thrown.

nhodges commented:
:+1: @frankweb thanks for the heads up. Fixed and we have a green build again.

gkalpak commented:
@nhodges, could you take care of the unresolved comments ?

gkalpak commented on docs/content/error/orderBy/notarray.ngdoc:
```diff
@@ -0,0 +1,53 @@
+             },
+      three: {
+             name: 'something 3'
+             },
```
Also, the indentation is funny for this object. TBH, I would change the example to something like:

```html
<!-- HTML -->
<label>
  Order by:
  <select ng-model="orderProp" ng-options="prop for prop in ['id', 'name']"></select>
</label>
<div ng-repeat="(key, value) in myObj | orderBy:orderProp">
  {{ key }} : {{ value }}
</div>
```

```js
/* JS */
angular.module('aModule', [])
  .controller('aController', function($scope) {
    var myObj = {
      one: {id: 1, name: 'Some thing'},
      two: {id: 2, name: 'Another thing'},
      three: {id: 3, name: 'A third thing'}
    };

    $scope.arrFromMyObj = Object.keys(myObj).map(function(key) {
      return myObj[key];
    });
  });
```

nhodges committed:
chore(orderBy): Update documentation based on code review.

nhodges commented:
@gkalpak You got it. Pushed.

nhodges commented:
@gkalpak @petebacondarwin @frankweb 

Anything else I can take care of?

frankweb commented:
@nhodges of course, I'll take care of this, do you mind if I do a new PR using your branch?

frankweb commented:
@nhodges Sorry for the last comment, I already deleted it, my comment was about other PR.

frankweb commented:
There is only one problem with the PR, I think all the commits should be made from one account, since you merged my branch this PR has commits with two accounts. Because of this, cla label says `no` @Narretz @petebacondarwin should @nhodges do another PR or what should he do?

nhodges commented:
Hey Francisco,

I think if you just sign the CLA should be good?

Made with thumbs.

> On Oct 19, 2015, at 5:38 PM, Francisco Hernández <notifications@github.com> wrote:
> 
> There is only one problem with the PR, I think all the commits should be made from one account, since you merged my branch this PR has commits with two accounts. Because of this, cla label says no @Narretz @petebacondarwin should @nhodges do another PR or what should he do?
> 
> —
> Reply to this email directly or view it on GitHub.
>

frankweb commented:
@nhodges I guess we should comment something to the googlebot, but I don't know exactly what.

frankweb commented:
@googlebot I'm okay with these commits.

frankweb commented:
@googlebot I authored these commits

nhodges commented:
@frankweb 

Hey Francisco, can you try clicking the link below and filling out the form?

http://code.google.com/legal/individual-cla-v1.0.html

Thanks!

frankweb commented:
@nhodges Of course, but I'm already signed in. I'll try again.

petebacondarwin commented:
@nhodges and @frankweb - as long as you have both signed the CLA then this is fine. The Google Bot always gets confused when a PR contains commits from more than one author.

frankweb commented:
@petebacondarwin Thank you it's good to know that. Do you need another change to this PR?

petebacondarwin commented on docs/content/error/orderBy/notarray.ngdoc:
````diff
@@ -0,0 +1,50 @@
+</div>
+```
+
+orderBy must be used with an array so a subset of items can be returned.
````
`orderBy must be used with an array-like input so a subset of items can be returned.`

petebacondarwin commented on docs/content/error/orderBy/notarray.ngdoc:
````diff
@@ -0,0 +1,50 @@
+```
+
+orderBy must be used with an array so a subset of items can be returned.
+The array can be initialized asynchronously and therefore null or undefined won't throw this error.
````
use backticks for `null` and `undefined`

petebacondarwin commented on docs/content/error/orderBy/notarray.ngdoc:
```diff
@@ -0,0 +1,50 @@
+orderBy must be used with an array so a subset of items can be returned.
+The array can be initialized asynchronously and therefore null or undefined won't throw this error.
+
+To orderBy an object by the value of its properties you can create your own array based on that object:
```
To use `orderBy` to order the properties of an object, you can create...

petebacondarwin commented on docs/content/error/orderBy/notarray.ngdoc:
```diff
@@ -0,0 +1,50 @@
+@ngdoc error
+@name orderBy:notarray
+@fullName Not an array
```
Value is not array-like

petebacondarwin commented on docs/content/error/orderBy/notarray.ngdoc:
```diff
@@ -0,0 +1,50 @@
+@fullName Not an array
+@description
+
+This error occurs when {@link ng.orderBy orderBy} is not used with an array:
```
is not passed an array-like value

petebacondarwin commented on test/ng/filter/orderBySpec.js:
```diff
@@ -8,6 +8,11 @@ describe('Filter: orderBy', function() {
+      expect(function() { orderBy({}); }).
+        toThrowMinErr('orderBy', 'notarray', 'Expected array but received: {}');
+    });
+
```
Please add a test that shows that an array-like (i.e. not a pure array) does not fail, unless there is already such a test in the spec.

petebacondarwin commented:
I have added a few minor comments.
Can you also update the documentation comment to make it clear that people can pass in array-like objects and that a non-array like object will result in an error?
Finally, since this is a breaking change, we need a BREAKING CHANGES block in one of the commits, if you could provide one.
We can squash all these commits together when we merge. It is likely to go into one commit. Can @nhodges and @frankweb agree on who would be the final author of this?

nhodges committed:
fix(orderBy): Modify docs and add test for array-like object.

BREAKING CHANGES:

These changes alter the behavior of AngularJS such that it will no
longer throw an error for non array values, as it will accept
array-like values as well. It will throw an error if no array-like
value is passed.

nhodges commented:
@petebacondarwin @frankweb 

I went ahead and took care of the requested changes to docs and added a test that takes in a strong and returns a sorted  array.

frankweb commented:
@nhodges Good! Please squash the commits, this is your pull request.

gkalpak commented on docs/content/error/orderBy/notarray.ngdoc:
```diff
@@ -0,0 +1,50 @@
+  Order by:
+  <select ng-model="orderProp" ng-options="prop for prop in ['id', 'name']"></select>
+</label>
+<div ng-repeat="(key, value) in myObj | orderBy:orderProp">
```
_myObj_  -->  _arrFromMyObj_

gkalpak commented on docs/content/error/orderBy/notarray.ngdoc:
````diff
@@ -0,0 +1,50 @@
+[toArrayFilter](https://github.com/petebacondarwin/angular-toArrayFilter):
+```html
+<input ng-model="search">
+<div ng-repeat="item in myObj | toArray:false | filter:search">
````
_filter:search_  -->  _orderBy:search_

gkalpak commented:
With https://github.com/angular/angular.js/pull/11719#issuecomment-150649515, I think we are good to go :smile:
(I left  couple of minor comments (typos), but I will take care of them while merging.)

Thx @nhodges and @frankweb (and @petebacondarwin) for working on this one !

gkalpak commented:
@nhodges, actually could you rebase your changes on master. It's been a long time since the first commit...

nhodges commented:
@gkalpak rebased. LMK if I can do anything else or if you want me to take care of the changes above.

gkalpak commented on src/ng/filter/orderBy.js:
```diff
@@ -177,7 +177,7 @@ orderByFilter.$inject = ['$parse'];
   return function(array, sortPredicate, reverseOrder) {
 
-    if (!(isArrayLike(array))) return array;
+    if (!(isArrayLike(array))) throw minErr('orderBy')('notarray', 'Expected array but received: {0}', array);
```
I was in the middle of merging when I realized there is an issue here: We need to allow `null`/`undefined` (in order to support asnc loading). 
This block needs to be modified (see [filter#L132-L138][1] for an example).

[1]: https://github.com/angular/angular.js/blob/3ca4ca463c97a2bbca09f0366d2a753a13751143/src/ng/filter/filter.js#L132-L138

frankweb replied:
@gkalpak  I guess something like 

``` javascript
if(!array) return array;

if (!(isArrayLike(array))) throw minErr('orderBy')('notarray', 'Expected array but received: {0}', array);
```

would be easier to read than the example.

This is just an idea,of course :) what do you think about that?

petebacondarwin replied:
I think that is about right and similar to what we do in other filters.

gkalpak replied:
In `filterFilter` we only allow null, undefined and array-like (and I think that should we do here as well). So it would be more like:

```js
if (array == null) return array;
```

gkalpak commented:
@nhodges, there seems to be one last little thing to do :smile:
Could you take care of https://github.com/angular/angular.js/pull/11719#discussion_r43314775 and add a test that verifies `null`/`undefined` do not result in an error ?

nhodges committed:
fix(orderBy): Return null/undefined as-is, add specs for this edge case.

nhodges commented:
@gkalpak Sure thing, please take a look at the latest changeset. Cheers!

gkalpak commented:
I updated the documentation (per https://github.com/angular/angular.js/pull/11719#issuecomment-150535425), made the change described in https://github.com/angular/angular.js/pull/11719#discussion_r43343003, fixed some typos in the error message and merged !

Well done everyone :smiley:
